### PR TITLE
fix asan failure for fold_conv_bn when conv is quantized

### LINKED ISSUE #42462: ASAN failure on module freezing
## 🐛 Bug

ASAN failures on master when calling module freezing on optimize_for_mobile on a module which has a child module.

For context, this is blocking full testing of https://github.com/pytorch/pytorch/pull/42041

## To Reproduce

1. patch this PR (https://github.com/pytorch/pytorch/pull/42461)
2. build with ASAN and run the test case, it will fail:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23
```

## Expected behavior

no ASAN failures

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42041 optimize_for_mobile: bring packed params to root module
* **#42461 fix asan failure for fold_conv_bn when conv is quantized**

Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

### CODE REVIEW

File: test/test_mobile_optimizer.py [hunk 1]
```diff
@@ -207,5 +208,45 @@ def get_lint_count_by_type(lint_type, module_lint_List):
...
+                x = self.dequant(x)
+                return x
+
+        def _quant_script_and_optimize(model):
+            model.qconfig = torch.quantization.get_default_qconfig('qnnpack')
+            torch.quantization.prepare(model, inplace=True)
+            model(torch.randn(4, 1, 4, 4))
+            torch.quantization.convert(model, inplace=True)
+            model = torch.jit.script(model)
+            model_optim = optimize_for_mobile(model)
```
Comment (vkuzo) [inline test/test_mobile_optimizer.py:246]: fails here

File: test/test_mobile_optimizer.py [hunk 2]
```diff
@@ -207,5 +208,46 @@ def get_lint_count_by_type(lint_type, module_lint_List):
...
+                return x
+
+        model = Parent()
+        model.qconfig = torch.quantization.get_default_qconfig('qnnpack')
+        torch.quantization.prepare(model, inplace=True)
+        model(torch.randn(4, 1, 4, 4))
+        torch.quantization.convert(model, inplace=True)
+        model = torch.jit.script(model)
+        # this line should not have ASAN failures
+        model_optim = optimize_for_mobile(model)
```
Comment (vkuzo) [inline test/test_mobile_optimizer.py:250]: if anyone has suggestions on something better to assert here (currently just a no-op asserting that no exceptions are raised), I'm flexible

File: torch/csrc/jit/passes/fold_conv_bn.cpp [hunk 3]
```diff
@@ -59,13 +59,21 @@ void replaceConvBiasWithGetAttr(Module& module) {
...
+  bool real_typename_ends_with_pattern_name =
+    real_typename.size() >= pattern_name.size() &&
       (0 ==
        real_typename.compare(
            real_typename.size() - pattern_name.size(),
            pattern_name.size(),
-           pattern_name))) {
+           pattern_name));
+  // we need to treat quantized modules in a special way
+  bool real_typename_contains_quantized =
```
Comment (vkuzo) [inline torch/csrc/jit/passes/fold_conv_bn.cpp:73]: if there is a better way to check for this which wouldn't catch people naming their floating point modules ".*quantized.*", I'm flexible

Comment (jerryzh168) [reply torch/csrc/jit/passes/fold_conv_bn.cpp:73]: can we do this: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16is_conv2d_moduleERKNS0_5MatchERKSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPNS0_5ValueESt4hashISA_ESt8equal_11899141

Comment (jerryzh168) [reply torch/csrc/jit/passes/fold_conv_bn.cpp:73]: we can add bias only when the module is conv{1,2,3}d

File: test/test_mobile_optimizer.py (hunk 2 above)
Comment (jerryzh168) [reply test/test_mobile_optimizer.py:250]: I think it's fine to omit the asserts if we just want to make sure it runs

**Review by jerryzh168 — APPROVED** [review]
Looks good, thanks!

**Comment by vkuzo** [comment]
resubmitted as https://github.com/pytorch/pytorch/pull/42739 due to a ghstack snafu

**Pull request closed without merging**